### PR TITLE
Add grouped patterns.

### PR DESCRIPTION
**Pull request by ehuss**
Stabilization PR: https://github.com/rust-lang/rust/pull/54497
Tracking issue: https://github.com/rust-lang/rust/issues/51087

### CODE REVIEW

File: src/patterns.md [hunk 1]
```diff
@@ -582,6 +583,25 @@ They are also used to [destructure](#destructuring) a tuple.
...
 
+## Grouped patterns
+
+> **<sup>Syntax</sup>**\
+> _GroupedPattern_ :\
+> &nbsp;&nbsp; `(` [_Pattern_] `)`
+
+Enclosing a pattern in parentheses can be used to explicitly control the
+precedence of compound patterns. For example, a reference pattern next to a
+range pattern such as `&0..=5` is ambiguous and is not allowed, but can be
```
Comment (Havvy) [inline src/patterns.md]: It's not ambiguous, and it *is* allowed. It means `(&0)..=5`. I wonder if we need a precedence table for patterns. But that'd probably be best done in its own PR.

Comment (ehuss) [reply src/patterns.md]: I'm a little confused.  Latest nightly says it is ambiguous:

```
error: the range pattern here has ambiguous interpretation
  --> src/main.rs:28:10
   |
28 |         &0..=30 => { println!("small") }
   |          ^^^^^^ help: add parentheses to clarify the precedence: `(0 ..=30)`

error: aborting due to previous error
```

I originally started writing a section on precedence, but I deleted it because I wasn't really clear what it could say (and I agree, it should probably be separate).  The only precedence rules I know of are this and paths vs identifiers (and that's not really a precedence but a shadowing rule).
